### ISSUE TITLE
The “bin” files RubyGems installs aren’t actually binary, improve naming 

### ISSUE THREAD

**Issue by bronzdoc** [rubygems/rubygems#2372]
See discussion in https://github.com/rubygems/rubygems/pull/2361

**Comment by tjouan** [comment]
I'll try to list the different names used in RubyGems and where they are used
later, but in the mean time I just found another relevant discussion here:
* https://github.com/rubygems/rubygems/pull/2099#issuecomment-347056279
* https://github.com/rubygems/rubygems/pull/2099#issuecomment-347069950

**Comment by MSP-Greg** [comment]
When I see bin/binary, I think compiled.  When I see exe/exec/executable, somewhat the same thing, with the additional issue of windows users are used to compiled apps having an .exe file extension.

Maybe 'CLI files' or 'CLI scripts'?  If a user isn't familiar with the term, Google 'CLI', and the first line is 'Command-line interface'.  I would hope that would be clear to people...

**Comment by segiddins** [comment]
We also call them `binstubs` in Bundler

**Comment by MSP-Greg** [comment]
It's also messy because Windows needs a bat/cmd file to launch the *nix style script from Ruby, and I think those are called binstubs in RG.

Maybe we should call them 'TFIYBEF', short for 'The Files In Your Bin or Exe Folders'.  Ok, been at a keyboard too long today...

**Comment by tjouan** [comment]
TL;DR:

* I would only change `--bindir` option *description*;
* I would remove reference to "stub" when it's about gem executable
  wrappers. I think of other things first ("test stub"…) and see it's
  used in `Gem::Specification` and `Gem::StubSpecification`. Referring
  to wrappers is less obvious for me with this term.


### `--bindir` option in many commands

  In my opinion the option name is correct and clear, this is the bin
directory, which is also the default value (`bin`) and referred as
such almost everywhere.
  I think "bin files" in description would be acceptable, because we
refer to their parent directory in this case, but "binary files" or
"binaries" is not correct IMHO and could be confused with extensions.


### Terminology used elsewhere

  I'll quote part of the doc for `Gem::Specification#executables` and
`Gem::Installer` which I think are clear:

    ##
    # Executables included in the gem.
    #
    # For example, the rake gem has rake as an executable. You don’t specify the
    # full path (as in bin/rake); all application-style files are expected to be
    # found in bindir.  These files must be executable Ruby files.  Files that
    # use bash or other interpreters will not work.
    #
    # Executables included may only be ruby scripts, not scripts for other
    # languages or compiled binaries.

    # Gem::Installer does the work of putting files in all the right places on the
    # filesystem including unpacking the gem into its gem dir, installing the
    # gemspec in the specifications dir, storing the cached gem in the cache dir,
    # and installing either wrappers or symlinks for executables.

  So in a gem, we have executable Ruby scripts, and we install either
an executable "wrapper" (which is also a Ruby script) or a symlink.
Those are the terms I personally prefer.


---

Notes (far from complete, might contain errors)
-----------------------------------------------

### executable and/or wrapper

  Used in option flags, help, code, code documentation, error
messages.

* `executables` in a gem specification.
* UI:
  + `gem install --[no-]wrappers --[no-]format-executable`;
  + `gem pristine --only-executables`;
  + `gem uninstall --[no-]executables --[no-]format-executable`;
  + `gem update --[no-]wrappers --[no-]format-executable`;
  + `setup.rb --[no-]format-executable`;
  + "For gems with executables ruby installs a wrapper file into the
    executable directory by default.";
  + "Use bin wrappers for executables";
  + "Only restore executables";
  + "Executables and scripts will remain installed."
  + "Rewrite executables with a shebang";
  + "RubyGems installed the following executables:";
  + `"can't find gem #{dep} with executable #{exec_name}"`;
  + `"#{spec.name}'s executable \"#{filename}\" conflicts with "`;
  + `ask_yes_no "#{question}\nOverwrite the executable?"`;
* code or code documentation:
  + "executable for gem";
  + "executable's path";
  + "gem executables".
  + `Gem.bin_path(name, exec_name = nil, *requirements)`;
  + `Gem.find_spec_for_exe`;
  + `Gem::Commands::SetupCommand` (`format_executable`);
  + `Gem::DependencyInstaller` (`format_executable`);
  + `Gem::Installer`;
  + `Gem::Specification`.

There is too much code examples for "executable" or just "exe", I gave
up at some point.
  In RubyGems code, I think "executable" alone may refer to the
wrapper, or occasionally to the interpreter (Ruby) specified in the
wrapper's shebang.


### bin and/or binary

  "bin" is mostly used to refer to a "gem's bin path".

  "binary files" is mostly used for extensions, with a few exceptions.

  "bin file" is sometimes used, but I understand it as "a file inside
the bin directory", not a "binary file").

  When referring to a directory, it can be either the one in a gem,
either the one in the installation directory.

* `bindir` in a gem specification.
* UI:
  + `gem install --bindir`;
  + `gem pristine --bindir`;
  + `gem uninstall --bindir`;
  + `gem update --bindir`.
  + "Use bin wrappers for executables";
  + "Directory where binary files are located".
* code or code documentation:
  + `Gem.bin_path`;
  + `Gem.bindir`;
  + `Gem.default_bindir` "The default directory for binaries";
  + `Gem::DependencyInstaller` (`bin_dir`);
  + `Gem::Installer` (`bin_dir` or `bindir`, `#generate_bin`,
    `#generate_bin_script`, `#generate_bin_symlink`, `#extract_bin`…).


### stub / bin stub

  "stub" is mostly used for "stubbed specification" which is unrelated
to executable wrappers/links if I understand correctly. Only a few
occurrences are about executables/wrappers/scripts.

* UI:
  + `gem pristine` command description: "all bin stubs for the gem";
  + `setup.rb --[no-]regenerate-binstubs` "Regenerate gem binstubs".
* code:
  + `Gem::Installer#windows_stub_script`.

**Comment by bronzdoc** [comment]
`I would only change --bindir option description`

so the description would be changed from:
```
Directory where binary files are 'located'
```
to: 
```
Directory where executables  are 'located'
```

Agree? @segiddins @tjouan @MSP-Greg

**Comment by tjouan** [comment]
On 2018-08-09 10:51:40 -0700, Luis Sagastume wrote:
> so the description would be changed from:
> ```
> Directory where binary files are 'located'
> ```
> to: 
> ```
> Directory where executables  are 'located'
> ```
> 
> Agree? @segiddins @tjouan @MSP-Greg 

  Yes. With only one space after "executables" of course.

  By the way there are no quotes around "located" currently, I think
it should stay that way.

  I also just noticed that there is a new line just before "located",
it's not directly relevant to this issue, but I wonder if it should be
removed? Description is not so long, and shorter than others on a
single line. For this last point, I notice other inconsistencies so
might open a new issue/PR to address this.
